chapter/background.md

```diff
@@ -0,0 +1,451 @@
+
+// simple bindings
+let name = <expr> in <expr>
+let func = fun arg => <expr> in <expr>
+
+// or with patterns
+let name @ { field, with_default = 2 } = <expr> in <expr>
+let func = fun arg @ { field, with_default = 2 } => 
+  <expr> in 
+  <expr>
+```
+
+Let bindings in their simplest form merely bind a name to a value expression and expose the name to the inner expression.
+Hence, the `Let` node contains the binding and links to both implementation and scope subtrees.
+The binding can be a simple name, a pattern or both by naming the pattern as shown in [@lst:nickel-let-binding].
+
+
+```{.nickel #lst:nickel-args-function caption="Parsed representation of functions with multiple arguments"}
+fun first second => first + second
+// ...is parsed as
+fun first =>
+  fun second => first + second
+```
+
+Functions in Nickel are curried lambda expressions.
+A function with multiple arguments gets broken down into nested single argument functions as seen in [@lst:nickel-args-function].
+Function argument name binding therefore looks the same as in `let` bindings.
+
+
+##### Meta Information
+
+One key feature of Nickel is its gradual typing system [ref again?], which implies that values can be explicitly typed.
+Complementing type information, it is possible to annotate values with contracts and additional metadata such as contracts, documentation, default values and merge priority using a special syntax as displayed in [@lst:nickel-meta].
+
+
+```{.nickel #lst:nickel-meta caption="Example of a static Nickel expression"}
+let Contract = { 
+  foo | Num 
+      | doc "I am foo",
+  hello | Str
+        | default = "world"
+}
+| doc "Just an example Contract"
+in 
+let value | #Contract = { foo = 9, }
+in value == { foo = 9, hello = "world", } 
+
+> true
+```
+
+Internally, the addition of annotations wraps the annotated term in a `MetaValue`, an additional tree node which describes its subtree. 
+The expression shown in [@lst:nickel-meta-typed] translates to the AST in [@fig:nickel-meta-typed].
+
+```{.nickel #lst:nickel-meta-typed caption="Example of a typed expression"}
+let x: Num = 5 in x
+```
+
+
+```{.graphviz #fig:nickel-meta-typed caption="AST of typed expression" height=4.5cm}
+strict digraph { 
+  graph [fontname = "Fira Code"];
+  node [fontname = "Fira Code"];
+  edge [fontname = "Fira Code"];
+
+  meta [label="MetaValue", color="green", shape="box"]
+  let [label = "Let('x')"]
+  num [label = "Num(5)"]
+  var [label = "Var('x')"]
+
+  meta -> let
+  let -> num
+  let -> var
+}
+```
+
+##### Nested Record Access
+
+Nickel supports both static and dynamic access to record fields.
+If the field name is statically known, the access is said to be *static* accordingly.
+Conversely, if the name requires evaluating a string from an expression the access is called *dynamic*.
+An example is given in [@lst:nickel-static-dynamic]
+
+```{.nickel #lst:nickel-static-dynamic caption="Examples for static and dynamic record access}
```

Comment:
```suggestion
```{.nickel #lst:nickel-static-dynamic caption="Examples for static and dynamic record access"}
```